src/linear_algebra/char_poly/coeff.lean

```diff
@@ -0,0 +1,191 @@
+(char_poly M).degree = fintype.card n :=
+begin
+  by_cases fintype.card n = 0, rw h, unfold char_poly, rw det_of_dim_zero, simpa,
+  rw ← sub_add_cancel (char_poly M) (∏ (i : n), (X - C (M i i))),
+  have h1 : (∏ (i : n), (X - C (M i i))).degree = fintype.card n,
+  { rw degree_eq_iff_nat_degree_eq_of_pos, swap, apply nat.pos_of_ne_zero h,
+    rw nat_degree_prod', simp_rw nat_degree_X_sub_C, unfold fintype.card, simp,
+    simp_rw (monic_X_sub_C _).leading_coeff, simp, },
+  rw degree_add_eq_of_degree_lt, exact h1, rw h1,
+  apply lt_trans (low_degree_char_poly_sub_diagonal M), rw with_bot.coe_lt_coe,
+  rw ← nat.pred_eq_sub_one, apply nat.pred_lt, apply h,
+end
+
+theorem nat_degree_char_poly_eq_dim [nontrivial R] (M: matrix n n R) :
+  (char_poly M).nat_degree = fintype.card n :=
+nat_degree_eq_of_degree_eq_some (degree_char_poly_eq_dim M)
+
+lemma char_poly_monic_of_nontrivial [nontrivial R] (M : matrix n n R) :
+  monic (char_poly M) :=
+begin
+  by_cases fintype.card n = 0, rw [char_poly, det_of_dim_zero h], apply monic_one,
+  have mon : (∏ (i : n), (X - C (M i i))).monic,
+  { apply monic_prod_of_monic univ (λ i : n, (X - C (M i i))), simp [monic_X_sub_C], },
+  rw ← sub_add_cancel (∏ (i : n), (X - C (M i i))) (char_poly M) at mon,
+  rw monic at *, rw leading_coeff_add_of_degree_lt at mon, rw ← mon,
+  rw degree_char_poly_eq_dim, rw ← neg_sub, rw degree_neg,
+  apply lt_trans (low_degree_char_poly_sub_diagonal M), rw with_bot.coe_lt_coe,
+  rw ← nat.pred_eq_sub_one, apply nat.pred_lt, apply h,
+end
+
+lemma char_poly_monic (M : matrix n n R) :
+  monic (char_poly M) :=
+begin
+  classical, by_cases h : nontrivial R,
+  { letI := h, apply char_poly_monic_of_nontrivial, },
+  { rw nontrivial_iff at h, push_neg at h, apply h, }
+end
+
+--shouldn't need these instances, but might need casework
+theorem trace_from_char_poly [nontrivial R] [nonempty n] (M: matrix n n R) :
+(matrix.trace n R R) M = -(char_poly M).coeff (fintype.card n - 1) :=
+begin
+  rw high_coeff_char_poly_eq_coeff_prod_diag, swap, refl,
+  rw [fintype.card, prod_X_sub_C_coeff_card_pred univ (λ i : n, M i i)], simp,
+  rw [← fintype.card, fintype.card_pos_iff], apply_instance,
+end
+
+lemma ring_hom_det {S : Type w} [comm_ring S] {M : matrix n n R} {f : R →+* S} :
+  f M.det = matrix.det (f.map_matrix M) :=
+by simp [matrix.det, f.map_sum, f.map_prod]
+
+lemma alg_hom_det {S : Type w} [comm_ring S] [algebra R S] {T : Type z} [comm_ring T] [algebra R T]
+  {M : matrix n n S} {f : S →ₐ[R] T} :
+  f M.det = matrix.det ((f : S →+* T).map_matrix M) :=
+by rw [← alg_hom.coe_to_ring_hom, ring_hom_det]
+
+lemma matrix.scalar.commute (r : R) (M : matrix n n R) : commute (scalar n r) M :=
+by simp [commute, semiconj_by]
+
+-- I feel like this should use polynomial.alg_hom_eval₂_algebra_map
+lemma eval_mat_poly_equiv (M : matrix n n (polynomial R)) (r : R) (i j : n) :
+  polynomial.eval r (M i j) = polynomial.eval ((scalar n) r) (mat_poly_equiv M) i j :=
```

Comment:
Use dot notation for `eval`?

If you reversed the LHS and RHS, the name would be good (although better as `mat_poly_equiv_eval`), but as is it's the wrong name.